File: python/src/ensoul/http.py
"""HTTP transport layer for the Ensoul SDK."""

from __future__ import annotations

import random
import time

import httpx

from ensoul._types import HeadersLike, QueryParams
from ensoul.auth import APIKeyAuth, AuthProvider, BearerAuth
from ensoul.config import API_VERSION, ClientConfig
from ensoul.errors import RateLimitError, raise_for_status
from ensoul.rate_limit import RateLimitTracker
from ensoul.streaming import AsyncSSEStream, SyncSSEStream

__all__ = [
    "SyncHTTPClient",
    "AsyncHTTPClient",
]

_RETRY_STATUS_CODES = {429, 500, 502, 503}
_SDK_USER_AGENT = "ensoul-python/0.1.0"


def _build_auth(config: ClientConfig) -> AuthProvider:
    if config.api_key is not None:
        return APIKeyAuth(config.api_key)
    if config.bearer_token is not None:
        return BearerAuth(access_token=config.bearer_token)
    # No credentials — requests will still go through (server decides)
    return APIKeyAuth("")


def _normalize_path(path: str) -> str:
    """Ensure path starts with /v1/. Handles paths with or without leading slash."""
    if path.startswith(f"/{API_VERSION}/") or path.startswith(f"{API_VERSION}/"):
        return path if path.startswith("/") else f"/{path}"
    path = path.lstrip("/")
    return f"/{API_VERSION}/{path}"


def _retry_wait(attempt: int, retry_after: float | None) -> float:
    """Compute seconds to wait before the next retry attempt.

    Uses the Retry-After header when available; otherwise exponential backoff
    with jitter: base * 2^attempt + random(0, 1).
    """
    if retry_after is not None and retry_after > 0:
        return retry_after
    base_wait = min(0.5 * (2 ** attempt), 30.0)
    jitter = random.uniform(0.0, 1.0)
    return base_wait + jitter


class SyncHTTPClient:
    """Synchronous HTTP transport client."""

    def __init__(self, config: ClientConfig) -> None:
        self._config = config
        self._auth: AuthProvider = _build_auth(config)
        self._rate_limiter = RateLimitTracker()

        default_headers: dict[str, str] = {
            "User-Agent": _SDK_USER_AGENT,
            "Accept": "application/json",
            **config.custom_headers,
        }
        self._client = httpx.Client(
            base_url=config.base_url,
            timeout=config.timeout,
            headers=default_headers,
            follow_redirects=True,
        )

    def request(
        self,
        method: str,
        path: str,
        *,
        json: dict | None = None,
        params: QueryParams | None = None,
        headers: HeadersLike | None = None,
        stream: bool = False,
    ) -> httpx.Response:
        """Make an authenticated HTTP request with retries and rate limit handling."""
        path = _normalize_path(path)
        last_exc: Exception | None = None

        for attempt in range(self._config.max_retries + 1):
            # Wait if rate limit is exhausted before sending
            should_wait, wait_seconds = self._rate_limiter.should_wait()
            if should_wait:
                time.sleep(wait_seconds)

            req_headers: dict[str, str] = {**self._auth.auth_headers()}
            if headers:
                req_headers.update(headers)

            try:
                if stream:
                    response = self._client.send(
                        self._client.build_request(
                            method,
                            path,
                            json=json,
                            params=params,
                            headers=req_headers,
                        ),
                        stream=True,
                    )
                else:
                    response = self._client.request(
                        method,
                        path,
                        json=json,
                        params=params,
                        headers=req_headers,
                    )

                self._rate_limiter.update(response)

                if response.status_code in _RETRY_STATUS_CODES and attempt < self._config.max_retries:
                    retry_after: float | None = None
                    if response.status_code == 429:
                        raw = response.headers.get("Retry-After")
                        if raw is not None:
                            try:
                                retry_after = float(raw)
                            except ValueError:
                                pass
                    wait = _retry_wait(attempt, retry_after)
                    time.sleep(wait)
                    continue

                raise_for_status(response)
                return response

            except RateLimitError:
                raise
            except httpx.TimeoutException as exc:
                last_exc = exc
                if attempt < self._config.max_retries:
                    time.sleep(_retry_wait(attempt, None))
                    continue
                raise
            except httpx.TransportError as exc:
                last_exc = exc
                if attempt < self._config.max_retries:
                    time.sleep(_retry_wait(attempt, None))
                    continue
                raise

        # Should not be reached; satisfy type checker
        if last_exc is not None:
            raise last_exc
        raise RuntimeError("Exhausted retries without a response")

    def get(self, path: str, *, params: QueryParams | None = None) -> httpx.Response:
        return self.request("GET", path, params=params)

    def get_raw(self, path: str, *, params: QueryParams | None = None) -> httpx.Response:
        """GET without /v1/ prefix normalization — for paths like /health."""
        req_headers: dict[str, str] = {**self._auth.auth_headers()}
        response = self._client.request("GET", path, params=params, headers=req_headers)
        raise_for_status(response)
        return response

    def post(self, path: str, *, json: dict | None = None, params: QueryParams | None = None) -> httpx.Response:
        return self.request("POST", path, json=json, params=params)

    def put(self, path: str, *, json: dict | None = None) -> httpx.Response:
        return self.request("PUT", path, json=json)

    def patch(self, path: str, *, json: dict | None = None) -> httpx.Response:
        return self.request("PATCH", path, json=json)

    def delete(self, path: str) -> httpx.Response:
        return self.request("DELETE", path)

    def stream_sse(
        self,
        method: str,
        path: str,
        *,
        json: dict | None = None,
        params: QueryParams | None = None,
    ) -> SyncSSEStream:
        """Make a streaming request and return a synchronous SSE stream."""
        response = self.request(method, path, json=json, params=params, stream=True)
        return SyncSSEStream(response)

    def close(self) -> None:
        """Close the underlying HTTP client and release connections."""
        self._client.close()

    def __enter__(self) -> SyncHTTPClient:
        return self

    def __exit__(self, *args: object) -> None:
        self.close()


class AsyncHTTPClient:
    """Asynchronous HTTP transport client."""

    def __init__(self, config: ClientConfig) -> None:
        self._config = config
        self._auth: AuthProvider = _build_auth(config)
        self._rate_limiter = RateLimitTracker()

        default_headers: dict[str, str] = {
            "User-Agent": _SDK_USER_AGENT,
            "Accept": "application/json",
            **config.custom_headers,
        }
        self._client = httpx.AsyncClient(
            base_url=config.base_url,
            timeout=config.timeout,
            headers=default_headers,
            follow_redirects=True,
        )

    async def request(
        self,
        method: str,
        path: str,
        *,
        json: dict | None = None,
        params: QueryParams | None = None,
        headers: HeadersLike | None = None,
        stream: bool = False,
    ) -> httpx.Response:
        """Make an authenticated async HTTP request with retries and rate limit handling."""
        import asyncio

        path = _normalize_path(path)
        last_exc: Exception | None = None

        for attempt in range(self._config.max_retries + 1):
            should_wait, wait_seconds = self._rate_limiter.should_wait()
            if should_wait:
                await asyncio.sleep(wait_seconds)

            req_headers: dict[str, str] = {**self._auth.auth_headers()}
            if headers:
                req_headers.update(headers)

            try:
                if stream:
                    response = await self._client.send(
                        self._client.build_request(
                            method,
                            path,
                            json=json,
                            params=params,
                            headers=req_headers,
                        ),
                        stream=True,
                    )
                else:
                    response = await self._client.request(
                        method,
                        path,
                        json=json,
                        params=params,
                        headers=req_headers,
                    )

                self._rate_limiter.update(response)

                if response.status_code in _RETRY_STATUS_CODES and attempt < self._config.max_retries:
                    retry_after: float | None = None
                    if response.status_code == 429:
                        raw = response.headers.get("Retry-After")
                        if raw is not None:
                            try:
                                retry_after = float(raw)
                            except ValueError:
                                pass
                    wait = _retry_wait(attempt, retry_after)
                    await asyncio.sleep(wait)
                    continue

                raise_for_status(response)
                return response

            except RateLimitError:
                raise
            except httpx.TimeoutException as exc:
                last_exc = exc
                if attempt < self._config.max_retries:
                    await asyncio.sleep(_retry_wait(attempt, None))
                    continue
                raise
            except httpx.TransportError as exc:
                last_exc = exc
                if attempt < self._config.max_retries:
                    await asyncio.sleep(_retry_wait(attempt, None))
                    continue
                raise

        if last_exc is not None:
            raise last_exc
        raise RuntimeError("Exhausted retries without a response")

    async def get(self, path: str, *, params: QueryParams | None = None) -> httpx.Response:
        return await self.request("GET", path, params=params)

    async def get_raw(self, path: str, *, params: QueryParams | None = None) -> httpx.Response:
        """GET without /v1/ prefix normalization — for paths like /health."""
        req_headers: dict[str, str] = {**self._auth.auth_headers()}
        response = await self._client.request("GET", path, params=params, headers=req_headers)
        raise_for_status(response)
        return response

    async def post(self, path: str, *, json: dict | None = None, params: QueryParams | None = None) -> httpx.Response:
        return await self.request("POST", path, json=json, params=params)

    async def put(self, path: str, *, json: dict | None = None) -> httpx.Response:
        return await self.request("PUT", path, json=json)

    async def patch(self, path: str, *, json: dict | None = None) -> httpx.Response:
        return await self.request("PATCH", path, json=json)

    async def delete(self, path: str) -> httpx.Response:
        return await self.request("DELETE", path)

    async def stream_sse(
        self,
        method: str,
        path: str,
        *,
        json: dict | None = None,
        params: QueryParams | None = None,
    ) -> AsyncSSEStream:
        """Make a streaming request and return an async SSE stream."""
        response = await self.request(method, path, json=json, params=params, stream=True)
        return AsyncSSEStream(response)

    async def close(self) -> None:
        """Close the underlying async HTTP client and release connections."""
        await self._client.aclose()

    async def __aenter__(self) -> AsyncHTTPClient:
        return self

    async def __aexit__(self, *args: object) -> None:
        await self.close()

File: typescript/src/http.ts
import { APIKeyAuth, type AuthProvider, BearerAuth } from "./auth.js";
import { API_VERSION, type ClientConfig, SDK_VERSION } from "./config.js";
import { RateLimitError, raiseForStatus } from "./errors.js";
import { RateLimitTracker } from "./rate-limit.js";

const RETRYABLE_STATUS_CODES = new Set([429, 500, 502, 503]);
// Methods safe to replay without creating duplicate side effects. POST/PATCH are
// excluded: replaying a POST that already reached the server (e.g. a 120s domain
// generation that read-timed-out the client) re-runs the inference and double-bills
// the caller.
const IDEMPOTENT_METHODS = new Set(["GET", "HEAD", "OPTIONS", "PUT", "DELETE"]);
// Status codes safe to retry even for non-idempotent methods, because the server did
// not process the request: 429 (rate limited) and 503 (unavailable). 500/502 are
// ambiguous for POST, so they are only retried for idempotent methods.
const NONIDEMPOTENT_RETRY_STATUS_CODES = new Set([429, 503]);

function shouldRetryStatus(method: string, status: number): boolean {
  if (!RETRYABLE_STATUS_CODES.has(status)) return false;
  if (IDEMPOTENT_METHODS.has(method.toUpperCase())) return true;
  return NONIDEMPOTENT_RETRY_STATUS_CODES.has(status);
}

function shouldRetryNetwork(method: string): boolean {
  return IDEMPOTENT_METHODS.has(method.toUpperCase());
}

export class HTTPClient {
  private readonly config: ClientConfig;
  private readonly auth: AuthProvider;
  private readonly rateLimitTracker: RateLimitTracker;

  constructor(config: ClientConfig) {
    this.config = config;
    this.rateLimitTracker = new RateLimitTracker();

    if (config.apiKey) {
      this.auth = new APIKeyAuth(config.apiKey);
    } else if (config.bearerToken) {
      this.auth = new BearerAuth(config.bearerToken);
    } else {
      this.auth = new APIKeyAuth("");
    }
  }

  private _defaultHeaders(): Record<string, string> {
    return {
      "User-Agent": `ensoul-typescript/${SDK_VERSION}`,
      Accept: "application/json",
      "Content-Type": "application/json",
      ...this.config.customHeaders,
      ...this.auth.authHeaders(),
    };
  }

  _normalizePath(path: string): string {
    const prefix = `/${API_VERSION}/`;
    if (path.startsWith(prefix) || path.startsWith(`/${API_VERSION}`)) {
      return path;
    }
    const stripped = path.startsWith("/") ? path.slice(1) : path;
    return `/${API_VERSION}/${stripped}`;
  }

  private _buildUrl(path: string): string {
    const base = this.config.baseUrl.replace(/\/$/, "");
    return `${base}${path}`;
  }

  private async _retryWait(attempt: number, retryAfter?: number): Promise<void> {
    let waitSeconds: number;
    if (retryAfter !== undefined && retryAfter > 0) {
      waitSeconds = retryAfter;
    } else {
      const backoff = Math.min(0.5 * Math.pow(2, attempt), 30);
      const jitter = Math.random() * 0.5;
      waitSeconds = backoff + jitter;
    }
    await new Promise<void>((resolve) => setTimeout(resolve, waitSeconds * 1000));
  }

  private async _request(
    method: string,
    path: string,
    options: {
      body?: unknown;
      query?: Record<string, unknown>;
      skipNormalize?: boolean;
    } = {}
  ): Promise<{ status: number; headers: Headers; json: () => Promise<Record<string, unknown>>; text: () => Promise<string> }> {
    const normalizedPath = options.skipNormalize ? path : this._normalizePath(path);
    let url = this._buildUrl(normalizedPath);

    if (options.query && Object.keys(options.query).length > 0) {
      const params = new URLSearchParams();
      for (const [key, value] of Object.entries(options.query)) {
        if (value !== undefined && value !== null) {
          params.set(key, String(value));
        }
      }
      const queryString = params.toString();
      if (queryString) {
        url = `${url}?${queryString}`;
      }
    }

    const headers = this._defaultHeaders();
    const requestInit: RequestInit = {
      method,
      headers,
    };

    if (options.body !== undefined) {
      requestInit.body = JSON.stringify(options.body);
    }

    let lastError: unknown;

    for (let attempt = 0; attempt <= this.config.maxRetries; attempt++) {
      const { wait, seconds } = this.rateLimitTracker.shouldWait();
      if (wait) {
        await this._retryWait(0, seconds);
      }

      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), this.config.timeout);
      requestInit.signal = controller.signal;

      try {
        const response = await fetch(url, requestInit);
        clearTimeout(timeoutId);

        this.rateLimitTracker.update(response.headers);

        if (shouldRetryStatus(method, response.status) && attempt < this.config.maxRetries) {
          let retryAfter: number | undefined;
          const retryAfterHeader = response.headers.get("Retry-After");
          if (retryAfterHeader) {
            retryAfter = parseInt(retryAfterHeader, 10);
            if (isNaN(retryAfter)) retryAfter = undefined;
          }
          await this._retryWait(attempt, retryAfter);
          continue;
        }

        if (!response.ok) {
          const body = (await response.json().catch(() => ({}))) as Record<string, unknown>;
          raiseForStatus(response.status, body, response.headers);
        }

        return {
          status: response.status,
          headers: response.headers,
          json: () => response.json() as Promise<Record<string, unknown>>,
          text: () => response.text(),
        };
      } catch (err) {
        clearTimeout(timeoutId);

        if (err instanceof RateLimitError) {
          if (attempt < this.config.maxRetries) {
            await this._retryWait(attempt, err.retryAfter);
            lastError = err;
            continue;
          }
          throw err;
        }

        if (err instanceof Error && err.name === "AbortError") {
          lastError = new Error(`Request timed out after ${this.config.timeout}ms`);
          if (shouldRetryNetwork(method) && attempt < this.config.maxRetries) {
            await this._retryWait(attempt);
            continue;
          }
          throw lastError;
        }

        // Network errors
        if (err instanceof TypeError && shouldRetryNetwork(method) && attempt < this.config.maxRetries) {
          lastError = err;
          await this._retryWait(attempt);
          continue;
        }

        throw err;
      }
    }

    throw lastError ?? new Error("Request failed after retries");
  }

  async get<T = Record<string, unknown>>(
    path: string,
    query?: Record<string, unknown>
  ): Promise<T> {
    const res = await this._request("GET", path, { query });
    return res.json() as Promise<T>;
  }

  async post<T = Record<string, unknown>>(
    path: string,
    body?: unknown,
    query?: Record<string, unknown>
  ): Promise<T> {
    const res = await this._request("POST", path, { body, query });
    return res.json() as Promise<T>;
  }

  async put<T = Record<string, unknown>>(
    path: string,
    body?: unknown,
    query?: Record<string, unknown>
  ): Promise<T> {
    const res = await this._request("PUT", path, { body, query });
    return res.json() as Promise<T>;
  }

  async patch<T = Record<string, unknown>>(
    path: string,
    body?: unknown,
    query?: Record<string, unknown>
  ): Promise<T> {
    const res = await this._request("PATCH", path, { body, query });
    return res.json() as Promise<T>;
  }

  async delete<T = Record<string, unknown>>(
    path: string,
    query?: Record<string, unknown>
  ): Promise<T> {
    const res = await this._request("DELETE", path, { query });
    if (res.status === 204) {
      return undefined as unknown as T;
    }
    return res.json() as Promise<T>;
  }

  async postForm<T = Record<string, unknown>>(
    path: string,
    data: Record<string, string>,
  ): Promise<T> {
    const normalizedPath = this._normalizePath(path);
    const url = this._buildUrl(normalizedPath);
    const headers = {
      ...this._defaultHeaders(),
      "Content-Type": "application/x-www-form-urlencoded",
    };
    const body = new URLSearchParams(data).toString();

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.config.timeout);

    try {
      const response = await fetch(url, {
        method: "POST",
        headers,
        body,
        signal: controller.signal,
      });
      clearTimeout(timeoutId);

      this.rateLimitTracker.update(response.headers);

      if (!response.ok) {
        const responseBody = (await response.json().catch(() => ({}))) as Record<string, unknown>;
        raiseForStatus(response.status, responseBody, response.headers);
      }

      return response.json() as Promise<T>;
    } catch (err) {
      clearTimeout(timeoutId);
      throw err;
    }
  }

  async getRaw(path: string, query?: Record<string, unknown>): Promise<Record<string, unknown>> {
    const res = await this._request("GET", path, { query, skipNormalize: true });
    return res.json();
  }

  /** GET an unversioned path returning the raw response body as text (e.g. a PEM key). */
  async getText(path: string, query?: Record<string, unknown>): Promise<string> {
    const res = await this._request("GET", path, { query, skipNormalize: true });
    return res.text();
  }

  async streamSSE(path: string, body?: unknown): Promise<Response> {
    const normalizedPath = this._normalizePath(path);
    const url = this._buildUrl(normalizedPath);

    const headers: Record<string, string> = {
      ...this._defaultHeaders(),
      Accept: "text/event-stream",
    };

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.config.timeout);

    try {
      const response = await fetch(url, {
        method: "POST",
        headers,
        body: body !== undefined ? JSON.stringify(body) : undefined,
        signal: controller.signal,
      });
      clearTimeout(timeoutId);

      if (!response.ok) {
        const responseBody = (await response.json().catch(() => ({}))) as Record<string, unknown>;
        raiseForStatus(response.status, responseBody, response.headers);
      }

      return response;
    } catch (err) {
      clearTimeout(timeoutId);
      throw err;
    }
  }

  close(): void {
    // No-op for fetch-based client; present for API parity with other SDK implementations
  }
}
